server/services/draftsService.ts

```diff
@@ -0,0 +1,116 @@
+import { RedisClient } from 'redis'
+import uuid from 'uuid'
+import { promisify } from 'util'
+
+export interface Draft<T> {
+  id: string
+  type: string
+  createdAt: Date
+  createdBy: { id: string }
+  updatedAt: Date
+  data: T
+}
+
+interface DraftDTO {
+  id: string
+  type: string
+  createdAt: string
+  createdBy: { id: string }
+  updatedAt: string
+  data: unknown
+}
+
+export interface Clock {
+  now: () => Date
+}
+
+export default class DraftsService {
+  constructor(
+    private readonly redis: RedisClient,
+    private readonly expiry: { seconds: number },
+    private readonly clock: Clock
+  ) {}
+
+  private redisKey(id: string): string {
+    return `draft:${id}`
+  }
+
+  private async fetchDraftDTO(id: string): Promise<DraftDTO | null> {
+    const response = await promisify(this.redis.get).bind(this.redis)(this.redisKey(id))
+
+    if (response === null) {
+      return null
+    }
+
+    return JSON.parse(response)
+  }
+
+  private async writeDraft<Data>(draft: Draft<Data>): Promise<void> {
+    const dto: DraftDTO = {
+      ...draft,
+      createdAt: draft.createdAt.toISOString(),
+      updatedAt: draft.updatedAt.toISOString(),
+    }
+
+    await promisify<string, string, string, number>(this.redis.set).bind(this.redis)(
+      this.redisKey(draft.id),
+      JSON.stringify(dto),
+      'EX', // Set the following expiry time on this key
+      this.expiry.seconds
+    )
+  }
+
+  async createDraft<Data>(type: string, initialData: Data, { userId }: { userId: string }): Promise<Draft<Data>> {
+    const now = this.clock.now()
+    const draft: Draft<Data> = {
+      id: uuid.v4(),
+      type,
+      data: initialData,
+      createdAt: now,
+      createdBy: { id: userId },
+      updatedAt: now,
+    }
+
+    await this.writeDraft(draft)
+
+    return draft
+  }
+
+  async fetchDraft<Data>(id: string, { userId }: { userId: string }): Promise<Draft<Data> | null> {
```

Comment:
Question: would we expect fetching a draft to also bump its expiry date? If we consider refreshing the page to constitute "activity" then perhaps yes. But on the other hand, perhaps having 24 hours from initiating the journey is plenty. (I didn't build the thing I'm asking about here yet just because it'd be a bit more fiddly and I wanted to check first.)

Reply:
Good question - I wonder if this is something we'd want to ask George G or Jonathan, as they might know more information about how users are completing these journeys.

I'd say 24 hours is probably fine if bumping the expiry date is going to be fiddly, unless we hear otherwise from the UR team.